default/AI/ShipDesignAI.py

```diff
@@ -0,0 +1,1573 @@
+    _rating_function()
+    _class_specific_filter()
+    _starting_guess()
+
+    For improved performance, maybe override _filling_algorithm() with a more specialised algorithm as well.
+    """
+    basename = "Default - Do not build"     # base design name
+    description = "Base class Ship type"    # design description
+    useful_part_classes = ALL_META_CLASSES  # only these parts are considered in the design process
+
+    filter_useful_parts = True              # removes any part not belonging to self.useful_part_classes
+    filter_inefficient_parts = False        # removes cost-inefficient parts (less capacity and less capacity/cost)
+
+    running_index = 1                       # running index used in the design name
+    running_index_needs_update = True       # To check if we need to update the running_index after a load.
+
+    def __init__(self):
+        """Make sure to call this constructor in each subclass."""
+        self.species = None         # name of the piloting species (string)
+        self.hull = None            # hull object (not hullname!)
+        self.partnames = []         # list of partnames (string)
+        self.parts = []             # list of actual part objects
+        self.attacks = {}           # {damage:count}
+        self.structure = 0
+        self.shields = 0
+        self.fuel = 0
+        self.speed = 0
+        self.stealth = 0
+        self.detection = 0
+        self.troops = 0
+        self.colonisation = -1      # -1 since 0 indicates an outpost (capacity = 0)
+        self.production_cost = 9999
+        self.production_time = 1
+        self.pid = -1               # planetID for checks on production cost if not LocationInvariant.
+        self.additional_specifications = AdditionalSpecifications()
+
+    def evaluate(self):
+        """ Return a rating for the design.
+
+        First, check if the minimum requirements are met. If so, return the _rating_function().
+        Otherwise, return a large negative number scaling with distance to the requirements.
+
+        :returns: float - rating of the current part/hull combo
+        """
+        self.update_stats()
+        ret_val = 0
+
+        # If we do not meet the requirements, we want to return a negative rating.
+        # However, we also need to make sure, that the closer we are to requirements,
+        # the better our rating is so the optimizing heuristic finds "the right way".
+        if self.fuel < self.additional_specifications.minimum_fuel:
+            ret_val += -1000 * (self.additional_specifications.minimum_fuel - self.fuel)
+        if self.speed < self.additional_specifications.minimum_speed:
+            ret_val += -1000 * (self.additional_specifications.minimum_speed - self.speed)
+        if self.structure < self.additional_specifications.minimum_structure:
+            ret_val += -1000 * (self.additional_specifications.minimum_structure - self.structure)
```

Comment:
If the design had a negative rating due to missing a pre-req, then I am inclined to think the only two reasonable courses of action for the client code will either be to decide not to build the ship, or to resubmit a design request with reduced requirements.  I don't think we should count on the penalty process necessary matching whatever requirement reductions the client code might choose to make.  It would certainly simplify things if we could assume that, so perhaps it deserves some more thought.  For this initial implementation though, I think we should just have the designer code return a None, and have the client code simply not build the ship in this case.  For now we could just leave a TODO in the client code about investigating using an extra cycle to reduce the requirements in this event.